From patchwork Mon Jul 23 00:47:34 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10539499
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id D19C417FD
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:26 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id BBAE828414
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:26 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id AE3C828478; Mon, 23 Jul 2018 00:48:26 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 5BDA528414
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:26 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1731221AbeGWBqX (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Sun, 22 Jul 2018 21:46:23 -0400
Received: from mail-pl0-f65.google.com ([209.85.160.65]:44770 "EHLO
        mail-pl0-f65.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1730429AbeGWBqV (ORCPT <rfc822;kvm@vger.kernel.org>);
        Sun, 22 Jul 2018 21:46:21 -0400
Received: by mail-pl0-f65.google.com with SMTP id m16-v6so7475988pls.11;
        Sun, 22 Jul 2018 17:47:46 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=DDljZacJxZG2OWdhiAliNYhkc3j//3d2iZVEuKfWFo4=;
        b=UCRhZQU57k9FZSbidw1iYEO504K0Gc8Pb41aE1noNpQ8MaG1ARdlfX2NZdVIOYNC7U
         OVDfvqomLizu3TcT5oGlYfsqNc0M7RU8xjjNdQ9vj+Sq5ia71lydd+uaWFpGJWtt3OgG
         bF1xYkX+ITqxnjyjyIlGEVS6XQHxmNgtLRNs+ALdBNih2UrpXwEqsBOwad7PKnmP1JSe
         J2591Kqif40I0QNT8zF4PWOb+tOr3Ohg3EBN4XjojSyp/mxiyqAbALgaXf3ydcsxZbzF
         k5Pg2/JIhgf1He9ZtPTRM3q7AnF3cHrxphlDX4QtnY1/ThQ/ZO8QyJ39U+Gv/MKkHVQT
         W/6A==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=DDljZacJxZG2OWdhiAliNYhkc3j//3d2iZVEuKfWFo4=;
        b=DcWCsupzO4I+Sl6cJhRyQAvwT31ZzK/2fefuIWVjsSlw9XIlSjdJMZNile7I3G1aIV
         l9yFwmIIe8UXkA3rSczfeOuzkpzPoxivx2yymutdnJngJm11y2p0TuZySBHUe6udc0aK
         YiYKGv1KjX+peA3CkxChmo7vX5l4QIv9UVAhxdFgaoyp6fD+d/E+uNkURm+5SByVDFcT
         Ane0btK8STJR5WDcgDHh+xtJt0lJBVAtq/m0kwSJMy/nT/zDS1Z5yBJq5jySENjsnCWE
         nIbL7jW4794QOYT/w4SB/Pe0/6ntAv7T8mmfujgNi/Z8A1hgr6UKkW0ZFssaBgCXS4Y4
         AaZw==
X-Gm-Message-State: AOUpUlHvN4LDmvGwizPixU8cIEjTYAdQQFoyonrBJnD3d6XFLdFkwN+8
        jdRkP/ayUwao7RotkJyG8zYyC7nj
X-Google-Smtp-Source: 
 AAOMgpdPsHtnaQIi3nPJS8akPpcqNR8dF6vrDzMwoPrQWwVhu8Mdj58SOLNtmD+8/6fZCCAE2KlitQ==
X-Received: by 2002:a17:902:14e:: with SMTP id
 72-v6mr2776525plb.299.1532306865990;
        Sun, 22 Jul 2018 17:47:45 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 e82-v6sm11646069pfk.87.2018.07.22.17.47.44
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Sun, 22 Jul 2018 17:47:45 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Vitaly Kuznetsov <vkuznets@redhat.com>
Subject: [PATCH v4 RESEND 1/6] KVM: X86: Add kvm hypervisor init time platform
 setup callback
Date: Mon, 23 Jul 2018 08:47:34 +0800
Message-Id: <1532306859-31166-2-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1532306859-31166-1-git-send-email-wanpengli@tencent.com>
References: <1532306859-31166-1-git-send-email-wanpengli@tencent.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

Add kvm hypervisor init time platform setup callback which
will be used to replace native apic hooks by pararvirtual
hooks.

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Cc: Vitaly Kuznetsov <vkuznets@redhat.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 arch/x86/kernel/kvm.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/arch/x86/kernel/kvm.c b/arch/x86/kernel/kvm.c
index 5b2300b..591bcf2 100644
--- a/arch/x86/kernel/kvm.c
+++ b/arch/x86/kernel/kvm.c
@@ -624,12 +624,22 @@ static uint32_t __init kvm_detect(void)
 	return kvm_cpuid_base();
 }
 
+static void __init kvm_apic_init(void)
+{
+}
+
+static void __init kvm_init_platform(void)
+{
+	x86_platform.apic_post_init = kvm_apic_init;
+}
+
 const __initconst struct hypervisor_x86 x86_hyper_kvm = {
 	.name			= "KVM",
 	.detect			= kvm_detect,
 	.type			= X86_HYPER_KVM,
 	.init.guest_late_init	= kvm_guest_init,
 	.init.x2apic_available	= kvm_para_available,
+	.init.init_platform	= kvm_init_platform,
 };
 
 static __init int activate_jump_labels(void)

From patchwork Mon Jul 23 00:47:35 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10539489
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 0741D14BC
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:47:57 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id E5FD626530
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:47:56 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id D90E927FA3; Mon, 23 Jul 2018 00:47:56 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 6434726530
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:47:55 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1731259AbeGWBqY (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Sun, 22 Jul 2018 21:46:24 -0400
Received: from mail-pl0-f68.google.com ([209.85.160.68]:43785 "EHLO
        mail-pl0-f68.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1730429AbeGWBqY (ORCPT <rfc822;kvm@vger.kernel.org>);
        Sun, 22 Jul 2018 21:46:24 -0400
Received: by mail-pl0-f68.google.com with SMTP id o7-v6so7476120plk.10;
        Sun, 22 Jul 2018 17:47:48 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=z2kprk72x7C+VlOw+kBql/qzp+eXhH1H7X2McD1Q2yc=;
        b=cy/X1JdhgA9ACJjcJPuZ41AoB5dlFddi39ifEh9Tj8ENjvBxQf0XbdvJVmz3hqKYsC
         2MKOQP4KqqnQ8PLs7nqlhxq6PmaPtYwmb+wSEUYPqK2YWnBCOo5xXPKEj3uv50OwwQtD
         ubXnxaKKEXJYlDXIy/Oq7ycVWbzcs3zBY8jZMNzA8fy7YhMNkYzFy90vwOFrDfaSsvVU
         p4mYBr93wFDvfYXOrX23E6hwYC4CZ2+XVMFxv3i4Zzwz8VA3+//IuK46qQoapQth8uV9
         UIn7BlsKsYMEV6PDhfqxx0RdU8CQgmPxnQYbtfcBwht5B5Y6iR+FfHnk+/WG3SvGnPRq
         AWBA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=z2kprk72x7C+VlOw+kBql/qzp+eXhH1H7X2McD1Q2yc=;
        b=povLN/12TA012zxRxXNVKO0ocHBazMyFCzzcin8/JXf91RBqGFpeMCtnCh0ryFAR9w
         0RRQEGC/jm7LkGbMF9SzKe9aIoy6ru3ixtOLglprgVy/gH9SEBBWn9FWN5Hc4AV0XfFa
         UrOQ6Tm2rSO+hC9pk1FXRylXYd/Tkp03h55vIL3n5PnWP4FOiqcyQLA4cp4KC8vwkiYg
         heRssqj0zJPN1yOuEgNdQf4PW6WLHVQgAYiXKm3cH08zbiaouNVp7mJthII2MbfWws46
         POb2Ly78J6qXrI4yXZYdZhOuzrFr08Qw51kPLj+j47UPlm7S2kE/BWyKhVwUTQs/L7XX
         zI4w==
X-Gm-Message-State: AOUpUlH2AwnsweYitbZW1ZwcEX8BfI2rtbi8sT2ijhDlHkbIIao3phqi
        +zSkEnE2hLLBR9uPdutwUpBbXlV6
X-Google-Smtp-Source: 
 AAOMgpdPXBNNH6KXCNikVba9IX6ozHp4AjFZ7Z9gYhC7e+EsWn3BJj6tMd6QrJgEysqtol6x4CJ+gQ==
X-Received: by 2002:a17:902:7891:: with SMTP id
 q17-v6mr11009764pll.186.1532306868264;
        Sun, 22 Jul 2018 17:47:48 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 e82-v6sm11646069pfk.87.2018.07.22.17.47.46
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Sun, 22 Jul 2018 17:47:47 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Vitaly Kuznetsov <vkuznets@redhat.com>
Subject: [PATCH v4 RESEND 2/6] KVM: X86: Implement PV IPIs in linux guest
Date: Mon, 23 Jul 2018 08:47:35 +0800
Message-Id: <1532306859-31166-3-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1532306859-31166-1-git-send-email-wanpengli@tencent.com>
References: <1532306859-31166-1-git-send-email-wanpengli@tencent.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

Implement paravirtual apic hooks to enable PV IPIs.

apic->send_IPI_mask
apic->send_IPI_mask_allbutself
apic->send_IPI_allbutself
apic->send_IPI_all

This patch lets a guest which sends multicast IPIs at most can handle 128 vCPUs 
per hypercall on 64-bit machines and 64 vCPUs per hypercall on 32-bit machines.

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Cc: Vitaly Kuznetsov <vkuznets@redhat.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 arch/x86/include/uapi/asm/kvm_para.h |  1 +
 arch/x86/kernel/kvm.c                | 86 ++++++++++++++++++++++++++++++++++++
 include/uapi/linux/kvm_para.h        |  1 +
 3 files changed, 88 insertions(+)

diff --git a/arch/x86/include/uapi/asm/kvm_para.h b/arch/x86/include/uapi/asm/kvm_para.h
index 0ede697..19980ec 100644
--- a/arch/x86/include/uapi/asm/kvm_para.h
+++ b/arch/x86/include/uapi/asm/kvm_para.h
@@ -28,6 +28,7 @@
 #define KVM_FEATURE_PV_UNHALT		7
 #define KVM_FEATURE_PV_TLB_FLUSH	9
 #define KVM_FEATURE_ASYNC_PF_VMEXIT	10
+#define KVM_FEATURE_PV_SEND_IPI	11
 
 #define KVM_HINTS_REALTIME      0
 
diff --git a/arch/x86/kernel/kvm.c b/arch/x86/kernel/kvm.c
index 591bcf2..eed6046 100644
--- a/arch/x86/kernel/kvm.c
+++ b/arch/x86/kernel/kvm.c
@@ -454,6 +454,88 @@ static void __init sev_map_percpu_data(void)
 }
 
 #ifdef CONFIG_SMP
+static void __send_ipi_mask(const struct cpumask *mask, int vector)
+{
+	unsigned long flags;
+	int cpu, apic_id, min = 0, max = 0;
+#ifdef CONFIG_X86_64
+	__uint128_t ipi_bitmap = 0;
+	int cluster_size = 128;
+#else
+	u64 ipi_bitmap = 0;
+	int cluster_size = 64;
+#endif
+
+	if (cpumask_empty(mask))
+		return;
+
+	local_irq_save(flags);
+
+	for_each_cpu(cpu, mask) {
+		apic_id = per_cpu(x86_cpu_to_apicid, cpu);
+		if (!ipi_bitmap) {
+			min = max = apic_id;
+		} else if (apic_id < min && max - apic_id < cluster_size) {
+			ipi_bitmap <<= min - apic_id;
+			min = apic_id;
+		} else if (apic_id < min + cluster_size) {
+			max = apic_id < max ? max : apic_id;
+		} else {
+			kvm_hypercall4(KVM_HC_SEND_IPI, (unsigned long)ipi_bitmap,
+				(unsigned long)(ipi_bitmap >> BITS_PER_LONG), min, vector);
+			min = max = apic_id;
+			ipi_bitmap = 0;
+		}
+		__set_bit(apic_id - min, (unsigned long *)&ipi_bitmap);
+	}
+
+	if (ipi_bitmap) {
+		kvm_hypercall4(KVM_HC_SEND_IPI, (unsigned long)ipi_bitmap,
+			(unsigned long)(ipi_bitmap >> BITS_PER_LONG), min, vector);
+	}
+
+	local_irq_restore(flags);
+}
+
+static void kvm_send_ipi_mask(const struct cpumask *mask, int vector)
+{
+	__send_ipi_mask(mask, vector);
+}
+
+static void kvm_send_ipi_mask_allbutself(const struct cpumask *mask, int vector)
+{
+	unsigned int this_cpu = smp_processor_id();
+	struct cpumask new_mask;
+	const struct cpumask *local_mask;
+
+	cpumask_copy(&new_mask, mask);
+	cpumask_clear_cpu(this_cpu, &new_mask);
+	local_mask = &new_mask;
+	__send_ipi_mask(local_mask, vector);
+}
+
+static void kvm_send_ipi_allbutself(int vector)
+{
+	kvm_send_ipi_mask_allbutself(cpu_online_mask, vector);
+}
+
+static void kvm_send_ipi_all(int vector)
+{
+	__send_ipi_mask(cpu_online_mask, vector);
+}
+
+/*
+ * Set the IPI entry points
+ */
+static void kvm_setup_pv_ipi(void)
+{
+	apic->send_IPI_mask = kvm_send_ipi_mask;
+	apic->send_IPI_mask_allbutself = kvm_send_ipi_mask_allbutself;
+	apic->send_IPI_allbutself = kvm_send_ipi_allbutself;
+	apic->send_IPI_all = kvm_send_ipi_all;
+	pr_info("KVM setup pv IPIs\n");
+}
+
 static void __init kvm_smp_prepare_cpus(unsigned int max_cpus)
 {
 	native_smp_prepare_cpus(max_cpus);
@@ -626,6 +708,10 @@ static uint32_t __init kvm_detect(void)
 
 static void __init kvm_apic_init(void)
 {
+#if defined(CONFIG_SMP)
+	if (kvm_para_has_feature(KVM_FEATURE_PV_SEND_IPI))
+		kvm_setup_pv_ipi();
+#endif
 }
 
 static void __init kvm_init_platform(void)
diff --git a/include/uapi/linux/kvm_para.h b/include/uapi/linux/kvm_para.h
index dcf629d..84f8fe3 100644
--- a/include/uapi/linux/kvm_para.h
+++ b/include/uapi/linux/kvm_para.h
@@ -26,6 +26,7 @@
 #define KVM_HC_MIPS_EXIT_VM		7
 #define KVM_HC_MIPS_CONSOLE_OUTPUT	8
 #define KVM_HC_CLOCK_PAIRING		9
+#define KVM_HC_SEND_IPI		10
 
 /*
  * hypercalls use architecture specific

From patchwork Mon Jul 23 00:47:36 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10539497
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id AD3EA14BC
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:17 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 990FB28414
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:17 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 8D11028478; Mon, 23 Jul 2018 00:48:17 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 2AB3228414
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:17 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1731297AbeGWBq1 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Sun, 22 Jul 2018 21:46:27 -0400
Received: from mail-pg1-f194.google.com ([209.85.215.194]:33744 "EHLO
        mail-pg1-f194.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1731287AbeGWBq0 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Sun, 22 Jul 2018 21:46:26 -0400
Received: by mail-pg1-f194.google.com with SMTP id r5-v6so10962341pgv.0;
        Sun, 22 Jul 2018 17:47:50 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=kw6302ch86pg258GLepkuZEk2pw2bUxOqBJi45V/0IA=;
        b=mxYMSNTjPaTuk9p6flXgdWta6SBmnXCoCTCepfsUIczcJWdbdvSP/MaPriFgwl85g+
         49JAZwnnqvOPm1/5V6kxWhZ2iBrrIUq5vLhNzcMvnklPqF/KvZ5R2I6vr6rppXY3tP/e
         gmJzaha8cxi+WyrhrOsS35aYdH8HM30k+IxCfKqfawK4wbpToPxys3rycZQXzgB71zU0
         HOj5e3sBwvH8LCup8G9DCP5otkQVNKJEHZyWF3KvYDV5ZSpPoqA2RF0FC1eoy2nVVLbb
         r1YGFJuap/5niBiosnkpRInmNOhZe6WnQaTYTBLrMsolYp/VHKNhlDWl9HDjUibkOaMi
         IJSA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=kw6302ch86pg258GLepkuZEk2pw2bUxOqBJi45V/0IA=;
        b=t2yVerLribMk7ljjiGucGk+TQl3WkoI5SG39KiWuLngbzMyIgahJaC+p58hRWyRwoT
         U3tatFvi9V7sWT1VvaG28HWxZWXF23/0NORdpSWWtGsD8cMQ86bwupAKJlzdnnPWjqyi
         KJBgG8NVw/KXqVM1hj+QTsbwYGjudlPjfZm2brACI5IQx7qUO0frwS5UwpdnuTEcxis3
         XcLYl9NbUWnUN0/bDvj4khi20fliUtW6sCoTEjSIkHCOYc5XaqQqAgY3FQLBb3/JSm/n
         SuFuv9vAzVnjC0MSC+Nnk7DafTI8mRpIk1kjsR/XKI0wb1ezEm7d/uKY1RZ5Uxyibw70
         OVXA==
X-Gm-Message-State: AOUpUlEUBdezh0Ey0WkKPNH0HhQn84lsHbWOaRC7/VXIPVVwQZSNv3Wx
        6dnSq3TX6RHKDpYgzpZD+lVzZlOH
X-Google-Smtp-Source: 
 AAOMgpdl0lxkAf0uWXMLGxqtRTG8IvSYsCxlLd48PhAZbAXOQnnJ38N1biNRAjQgL6VfwGIYCxZyXw==
X-Received: by 2002:a62:d98f:: with SMTP id
 b15-v6mr11035185pfl.1.1532306870251;
        Sun, 22 Jul 2018 17:47:50 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 e82-v6sm11646069pfk.87.2018.07.22.17.47.48
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Sun, 22 Jul 2018 17:47:49 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Vitaly Kuznetsov <vkuznets@redhat.com>
Subject: [PATCH v4 RESEND 3/6] KVM: X86: Fallback to original apic hooks when
 bad happens
Date: Mon, 23 Jul 2018 08:47:36 +0800
Message-Id: <1532306859-31166-4-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1532306859-31166-1-git-send-email-wanpengli@tencent.com>
References: <1532306859-31166-1-git-send-email-wanpengli@tencent.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

Fallback to original apic hooks when unlikely kvm fails to add the 
pending IRQ to lapic.

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Cc: Vitaly Kuznetsov <vkuznets@redhat.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 arch/x86/kernel/kvm.c | 20 +++++++++++++-------
 1 file changed, 13 insertions(+), 7 deletions(-)

diff --git a/arch/x86/kernel/kvm.c b/arch/x86/kernel/kvm.c
index eed6046..57eb4a2 100644
--- a/arch/x86/kernel/kvm.c
+++ b/arch/x86/kernel/kvm.c
@@ -47,6 +47,7 @@
 #include <asm/hypervisor.h>
 #include <asm/kvm_guest.h>
 
+static struct apic orig_apic;
 static int kvmapf = 1;
 
 static int __init parse_no_kvmapf(char *arg)
@@ -454,10 +455,10 @@ static void __init sev_map_percpu_data(void)
 }
 
 #ifdef CONFIG_SMP
-static void __send_ipi_mask(const struct cpumask *mask, int vector)
+static int __send_ipi_mask(const struct cpumask *mask, int vector)
 {
 	unsigned long flags;
-	int cpu, apic_id, min = 0, max = 0;
+	int cpu, apic_id, min = 0, max = 0, ret = 0;
 #ifdef CONFIG_X86_64
 	__uint128_t ipi_bitmap = 0;
 	int cluster_size = 128;
@@ -467,7 +468,7 @@ static void __send_ipi_mask(const struct cpumask *mask, int vector)
 #endif
 
 	if (cpumask_empty(mask))
-		return;
+		return 0;
 
 	local_irq_save(flags);
 
@@ -481,7 +482,7 @@ static void __send_ipi_mask(const struct cpumask *mask, int vector)
 		} else if (apic_id < min + cluster_size) {
 			max = apic_id < max ? max : apic_id;
 		} else {
-			kvm_hypercall4(KVM_HC_SEND_IPI, (unsigned long)ipi_bitmap,
+			ret = kvm_hypercall4(KVM_HC_SEND_IPI, (unsigned long)ipi_bitmap,
 				(unsigned long)(ipi_bitmap >> BITS_PER_LONG), min, vector);
 			min = max = apic_id;
 			ipi_bitmap = 0;
@@ -490,11 +491,12 @@ static void __send_ipi_mask(const struct cpumask *mask, int vector)
 	}
 
 	if (ipi_bitmap) {
-		kvm_hypercall4(KVM_HC_SEND_IPI, (unsigned long)ipi_bitmap,
+		ret = kvm_hypercall4(KVM_HC_SEND_IPI, (unsigned long)ipi_bitmap,
 			(unsigned long)(ipi_bitmap >> BITS_PER_LONG), min, vector);
 	}
 
 	local_irq_restore(flags);
+	return ret;
 }
 
 static void kvm_send_ipi_mask(const struct cpumask *mask, int vector)
@@ -511,7 +513,8 @@ static void kvm_send_ipi_mask_allbutself(const struct cpumask *mask, int vector)
 	cpumask_copy(&new_mask, mask);
 	cpumask_clear_cpu(this_cpu, &new_mask);
 	local_mask = &new_mask;
-	__send_ipi_mask(local_mask, vector);
+	if (__send_ipi_mask(local_mask, vector))
+		orig_apic.send_IPI_mask_allbutself(mask, vector);
 }
 
 static void kvm_send_ipi_allbutself(int vector)
@@ -521,7 +524,8 @@ static void kvm_send_ipi_allbutself(int vector)
 
 static void kvm_send_ipi_all(int vector)
 {
-	__send_ipi_mask(cpu_online_mask, vector);
+	if (__send_ipi_mask(cpu_online_mask, vector))
+		orig_apic.send_IPI_all(vector);
 }
 
 /*
@@ -529,6 +533,8 @@ static void kvm_send_ipi_all(int vector)
  */
 static void kvm_setup_pv_ipi(void)
 {
+	orig_apic = *apic;
+
 	apic->send_IPI_mask = kvm_send_ipi_mask;
 	apic->send_IPI_mask_allbutself = kvm_send_ipi_mask_allbutself;
 	apic->send_IPI_allbutself = kvm_send_ipi_allbutself;

From patchwork Mon Jul 23 00:47:37 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10539495
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 8356517FD
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:13 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 6E5A128414
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:13 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 625A128478; Mon, 23 Jul 2018 00:48:13 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id CAEFE28414
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:12 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S2387696AbeGWBq3 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Sun, 22 Jul 2018 21:46:29 -0400
Received: from mail-pg1-f195.google.com ([209.85.215.195]:43668 "EHLO
        mail-pg1-f195.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1731287AbeGWBq2 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Sun, 22 Jul 2018 21:46:28 -0400
Received: by mail-pg1-f195.google.com with SMTP id v13-v6so10958557pgr.10;
        Sun, 22 Jul 2018 17:47:52 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=V7zwJ0ywobsjSZWTW9QqXuo56/lwfNePcwyAd8m7TDU=;
        b=KTfNN6xEISypEnZ5qj5Rw5zpUOGpoP/YkcmLiOXfSHlPM6qJuaMOz4xAENHZcT3kgR
         o56jlseOxCKLo0kv4mxBacVbDcZfn5C9MlR/XPzHeF0YEvhcVJ6Ry9p+dkGUujHrJ8cC
         /fihnOOGg8DI3BiwOXHBCyOVDtQXPeVA9DcKODQQPdVEldDTQDu71OO2hXfGhkBeYNWQ
         G5Mc4UtVUwXeNtHP4M2tb6T9lpklmjsIaKl4qzvi2Phg/VvorHO4pKbV+q8AOCs94+oN
         BvX+3QzdbzhOYp6ejvp4S+mlHSZS24K4ZYAZ/PNasiIrxQmVhajvXpylU9j/euXLV4Be
         icqA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=V7zwJ0ywobsjSZWTW9QqXuo56/lwfNePcwyAd8m7TDU=;
        b=Vq7Td5VQ+4D3NC0gX+hyk52l4ihEBeudbSyY9iV+6C5/Z36SNPcB7+m7i7rclQvdjn
         yO+XJ+t5ow4t7zqS8H0X8SJ5AlhX5MfmkmGIwAlLQmJS/Em45Qt0BRQ3pox9UV0ko3MK
         1o/1igSFeI7XBDLf8gFDEQAqxXncTCsMqVSnI1gnUsiOtSrQBH+gHWXr8OWz54z+94M+
         jWuwIO+dYAO/f/bhdP1lBC+DbEbDFIhXhFEXz3I27CQR2F5fvjJnj+favokxMjwi4QFa
         x6HxJCxOBtoZnD91cT2mgYp0pheB7TkLrE+0CDJosg3A29d+iKeVUr+9LFFpMTvLFBdR
         L2AQ==
X-Gm-Message-State: AOUpUlG9q6YNGBfGs8ckzj4ejuxznCOhuM4azXEO9WEqg0wbf6LJz8uJ
        VtMIkv1xINVS+MgfrKD4hJAW1u4L
X-Google-Smtp-Source: 
 AAOMgpcbgdm8ieY59I90LZu4qB4dl2qufXauxlNkEhZuuMDAxPAisewbJhakHAyR+ILWIs8y74PIqA==
X-Received: by 2002:a62:384:: with SMTP id
 126-v6mr10995711pfd.11.1532306872328;
        Sun, 22 Jul 2018 17:47:52 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 e82-v6sm11646069pfk.87.2018.07.22.17.47.50
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Sun, 22 Jul 2018 17:47:51 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Vitaly Kuznetsov <vkuznets@redhat.com>
Subject: [PATCH v4 RESEND 4/6] KVM: X86: Implement PV IPIs send hypercall
Date: Mon, 23 Jul 2018 08:47:37 +0800
Message-Id: <1532306859-31166-5-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1532306859-31166-1-git-send-email-wanpengli@tencent.com>
References: <1532306859-31166-1-git-send-email-wanpengli@tencent.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

Using hypercall to send IPIs by one vmexit instead of one by one for
xAPIC/x2APIC physical mode and one vmexit per-cluster for x2APIC cluster 
mode. Intel guest can enter x2apic cluster mode when interrupt remmaping 
is enabled in qemu, however, latest AMD EPYC still just supports xapic 
mode which can get great improvement by Exit-less IPIs. This patchset 
lets a guest which sends multicast IPIs at most can handle 128 vCPUs per 
hypercall on 64-bit machines and 64 vCPUs per hypercall on 32-bit machines.

Hardware: Xeon Skylake 2.5GHz, 2 sockets, 40 cores, 80 threads, the VM 
is 80 vCPUs, IPI microbenchmark(https://lkml.org/lkml/2017/12/19/141):

x2apic cluster mode, vanilla

 Dry-run:                         0,            2392199 ns
 Self-IPI:                  6907514,           15027589 ns
 Normal IPI:              223910476,          251301666 ns
 Broadcast IPI:                   0,         9282161150 ns
 Broadcast lock:                  0,         8812934104 ns

x2apic cluster mode, pv-ipi 

 Dry-run:                         0,            2449341 ns
 Self-IPI:                  6720360,           15028732 ns
 Normal IPI:              228643307,          255708477 ns
 Broadcast IPI:                   0,         7572293590 ns  => 22% performance boost 
 Broadcast lock:                  0,         8316124651 ns

x2apic physical mode, vanilla

 Dry-run:                         0,            3135933 ns
 Self-IPI:                  8572670,           17901757 ns
 Normal IPI:              226444334,          255421709 ns
 Broadcast IPI:                   0,        19845070887 ns
 Broadcast lock:                  0,        19827383656 ns

x2apic physical mode, pv-ipi

 Dry-run:                         0,            2446381 ns
 Self-IPI:                  6788217,           15021056 ns
 Normal IPI:              219454441,          249583458 ns
 Broadcast IPI:                   0,         7806540019 ns  => 154% performance boost 
 Broadcast lock:                  0,         9143618799 ns

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Cc: Vitaly Kuznetsov <vkuznets@redhat.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 Documentation/virtual/kvm/hypercalls.txt | 17 ++++++++++++++
 arch/x86/kvm/x86.c                       | 38 ++++++++++++++++++++++++++++++++
 2 files changed, 55 insertions(+)

diff --git a/Documentation/virtual/kvm/hypercalls.txt b/Documentation/virtual/kvm/hypercalls.txt
index a890529..912b877 100644
--- a/Documentation/virtual/kvm/hypercalls.txt
+++ b/Documentation/virtual/kvm/hypercalls.txt
@@ -121,3 +121,20 @@ compute the CLOCK_REALTIME for its clock, at the same instant.
 
 Returns KVM_EOPNOTSUPP if the host does not use TSC clocksource,
 or if clock type is different than KVM_CLOCK_PAIRING_WALLCLOCK.
+
+6. KVM_HC_SEND_IPI
+------------------------
+Architecture: x86
+Status: active
+Purpose: Hypercall used to send IPIs.
+
+a0: ipi_bitmap low 64 bits
+a1: ipi_bitmap high 64 bits
+a2: the lowest APIC ID in bitmap
+a3: APIC ICR
+
+The hypercall lets a guest send multicast IPIs at most can handle
+128 vCPUs per hypercall on 64-bit machines and 64 vCPUs per hypercall
+on 32-bit machines.
+
+Returns 0 if successfully delivery the IPIs and 1 if discarded.
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 2b812b3..016c7e2 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -6691,6 +6691,41 @@ static void kvm_pv_kick_cpu_op(struct kvm *kvm, unsigned long flags, int apicid)
 	kvm_irq_delivery_to_apic(kvm, NULL, &lapic_irq, NULL);
 }
 
+/*
+ * Return 0 if successfully added and 1 if discarded.
+ */
+static int kvm_pv_send_ipi(struct kvm *kvm, unsigned long ipi_bitmap_low,
+		unsigned long ipi_bitmap_high, int min, int vector, int op_64_bit)
+{
+	int i;
+	struct kvm_apic_map *map;
+	struct kvm_vcpu *vcpu;
+	struct kvm_lapic_irq irq = {
+		.delivery_mode = APIC_DM_FIXED,
+		.vector = vector,
+	};
+
+	rcu_read_lock();
+	map = rcu_dereference(kvm->arch.apic_map);
+
+	for_each_set_bit(i, &ipi_bitmap_low, BITS_PER_LONG) {
+		vcpu = map->phys_map[min + i]->vcpu;
+		if (!kvm_apic_set_irq(vcpu, &irq, NULL))
+			return 1;
+	}
+
+	if (op_64_bit) {
+		for_each_set_bit(i, &ipi_bitmap_high, BITS_PER_LONG) {
+			vcpu = map->phys_map[min + i + BITS_PER_LONG]->vcpu;
+			if (!kvm_apic_set_irq(vcpu, &irq, NULL))
+				return 1;
+		}
+	}
+
+	rcu_read_unlock();
+	return 0;
+}
+
 void kvm_vcpu_deactivate_apicv(struct kvm_vcpu *vcpu)
 {
 	vcpu->arch.apicv_active = false;
@@ -6739,6 +6774,9 @@ int kvm_emulate_hypercall(struct kvm_vcpu *vcpu)
 	case KVM_HC_CLOCK_PAIRING:
 		ret = kvm_pv_clock_pairing(vcpu, a0, a1);
 		break;
+	case KVM_HC_SEND_IPI:
+		ret = kvm_pv_send_ipi(vcpu->kvm, a0, a1, a2, a3, op_64_bit);
+		break;
 #endif
 	default:
 		ret = -KVM_ENOSYS;

From patchwork Mon Jul 23 00:47:38 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10539493
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 4E9C017FD
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:08 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 3A56F28414
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:08 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 2ED8328478; Mon, 23 Jul 2018 00:48:08 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id C076628414
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:07 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S2387771AbeGWBqc (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Sun, 22 Jul 2018 21:46:32 -0400
Received: from mail-pg1-f196.google.com ([209.85.215.196]:45398 "EHLO
        mail-pg1-f196.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S2387749AbeGWBqa (ORCPT <rfc822;kvm@vger.kernel.org>);
        Sun, 22 Jul 2018 21:46:30 -0400
Received: by mail-pg1-f196.google.com with SMTP id f1-v6so10950874pgq.12;
        Sun, 22 Jul 2018 17:47:54 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=zAj0wpkzMgjSjcinyt8YwVCTAp82w+xoQBngFo020p8=;
        b=K8tzXOGHCF4RPc6cyLqAByho/ZV8tQ4pJr3PWDE9S05zjQpFGeUuWlCNdve1i+T6Q1
         no1jVhBMbnGW2HoFyG9NaUdvYeK0L9UV8mXfhkMBxayY6My7d/B30ZElo9tcpXg1toQf
         VUa9WzVnLgxfv4acSvavjRpsKGRNrMODsS8DTksWOOnvu9Cz2xRb8qh6xHR0hNfNpGHb
         qJEfjYYLcLgdCtH5FKzvh266g/IasOURPpkufuTShB2d7NJPQw1kM8qt4EpOiFrIxVT5
         QRQfWQ0iYLlD4nmJ/q2C/s4BpoBAg0CEZjYB7dpQGnxfeTwx0K2jw58U8FKxaFa8DTRK
         VgNA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=zAj0wpkzMgjSjcinyt8YwVCTAp82w+xoQBngFo020p8=;
        b=uZG66QeRje6P1oDh07u2iD8pfRFw8xMlJEMBE0F3k5Jvxw/mOd3DuF9KvhaaiM05Cw
         E4rE6PHHG13tz2lztjc6t/UOrdbOMiav0i6HzgEcixDV73l0qToqRP6GNdHzy4+t9tBR
         Qa2gm5/uFFec0lEmC+1IrLoThHwx/iVz+kxbAqCB7kMk8YrBRH82EXL/ESb1IEI8YAX1
         +qcKexdq1dDKo8DezU/S41je9qhBd1exoWLmPveoIC0s+x7n6IKUCeKro7hglaoUAyUZ
         yX93yB2CyJCuYkmY3k+KWFcTbZ6dIu2PPv5I/UCMyk9zrmCL5Uh/XWefGff7OszAqRTj
         ZuIA==
X-Gm-Message-State: AOUpUlFT/wskYLx0LzrsF0N2ass6Mrt8yYuoNnPReITwB97H4Q1pyneB
        80c2TGIrgPC/iSJzfXcW6m2fMf6o
X-Google-Smtp-Source: 
 AAOMgpeNWKofoSf3wQ8f/ETTffK+ohmzn38jD/oHJ6Ve54LRWNt6D/RSJkkE1e9Nico3LHDSFLa8CA==
X-Received: by 2002:a63:d15:: with SMTP id
 c21-v6mr10300734pgl.322.1532306874342;
        Sun, 22 Jul 2018 17:47:54 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 e82-v6sm11646069pfk.87.2018.07.22.17.47.52
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Sun, 22 Jul 2018 17:47:53 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Vitaly Kuznetsov <vkuznets@redhat.com>
Subject: [PATCH v4 RESEND 5/6] KVM: X86: Add NMI support to PV IPIs
Date: Mon, 23 Jul 2018 08:47:38 +0800
Message-Id: <1532306859-31166-6-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1532306859-31166-1-git-send-email-wanpengli@tencent.com>
References: <1532306859-31166-1-git-send-email-wanpengli@tencent.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

The NMI delivery mode of ICR is used to deliver an NMI to the processor, 
and the vector information is ignored.

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Cc: Vitaly Kuznetsov <vkuznets@redhat.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 arch/x86/kernel/kvm.c | 15 ++++++++++++---
 arch/x86/kvm/x86.c    | 16 +++++++++++-----
 2 files changed, 23 insertions(+), 8 deletions(-)

diff --git a/arch/x86/kernel/kvm.c b/arch/x86/kernel/kvm.c
index 57eb4a2..3456531 100644
--- a/arch/x86/kernel/kvm.c
+++ b/arch/x86/kernel/kvm.c
@@ -458,7 +458,7 @@ static void __init sev_map_percpu_data(void)
 static int __send_ipi_mask(const struct cpumask *mask, int vector)
 {
 	unsigned long flags;
-	int cpu, apic_id, min = 0, max = 0, ret = 0;
+	int cpu, apic_id, min = 0, max = 0, ret = 0, icr = 0;
 #ifdef CONFIG_X86_64
 	__uint128_t ipi_bitmap = 0;
 	int cluster_size = 128;
@@ -472,6 +472,15 @@ static int __send_ipi_mask(const struct cpumask *mask, int vector)
 
 	local_irq_save(flags);
 
+	switch (vector) {
+	default:
+		icr = APIC_DM_FIXED | vector;
+		break;
+	case NMI_VECTOR:
+		icr = APIC_DM_NMI;
+		break;
+	}
+
 	for_each_cpu(cpu, mask) {
 		apic_id = per_cpu(x86_cpu_to_apicid, cpu);
 		if (!ipi_bitmap) {
@@ -483,7 +492,7 @@ static int __send_ipi_mask(const struct cpumask *mask, int vector)
 			max = apic_id < max ? max : apic_id;
 		} else {
 			ret = kvm_hypercall4(KVM_HC_SEND_IPI, (unsigned long)ipi_bitmap,
-				(unsigned long)(ipi_bitmap >> BITS_PER_LONG), min, vector);
+				(unsigned long)(ipi_bitmap >> BITS_PER_LONG), min, icr);
 			min = max = apic_id;
 			ipi_bitmap = 0;
 		}
@@ -492,7 +501,7 @@ static int __send_ipi_mask(const struct cpumask *mask, int vector)
 
 	if (ipi_bitmap) {
 		ret = kvm_hypercall4(KVM_HC_SEND_IPI, (unsigned long)ipi_bitmap,
-			(unsigned long)(ipi_bitmap >> BITS_PER_LONG), min, vector);
+			(unsigned long)(ipi_bitmap >> BITS_PER_LONG), min, icr);
 	}
 
 	local_irq_restore(flags);
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 016c7e2..c9dbc2c 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -6695,15 +6695,21 @@ static void kvm_pv_kick_cpu_op(struct kvm *kvm, unsigned long flags, int apicid)
  * Return 0 if successfully added and 1 if discarded.
  */
 static int kvm_pv_send_ipi(struct kvm *kvm, unsigned long ipi_bitmap_low,
-		unsigned long ipi_bitmap_high, int min, int vector, int op_64_bit)
+		unsigned long ipi_bitmap_high, int min, unsigned long icr, int op_64_bit)
 {
 	int i;
 	struct kvm_apic_map *map;
 	struct kvm_vcpu *vcpu;
-	struct kvm_lapic_irq irq = {
-		.delivery_mode = APIC_DM_FIXED,
-		.vector = vector,
-	};
+	struct kvm_lapic_irq irq = {0};
+
+	switch (icr & APIC_VECTOR_MASK) {
+	default:
+		irq.vector = icr & APIC_VECTOR_MASK;
+		break;
+	case NMI_VECTOR:
+		break;
+	}
+	irq.delivery_mode = icr & APIC_MODE_MASK;
 
 	rcu_read_lock();
 	map = rcu_dereference(kvm->arch.apic_map);

From patchwork Mon Jul 23 00:47:39 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10539491
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 59CB514BC
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:05 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 4690A27FA3
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:05 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 3B01D283B2; Mon, 23 Jul 2018 00:48:05 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id C6D7427FA3
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 23 Jul 2018 00:48:04 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S2387787AbeGWBqc (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Sun, 22 Jul 2018 21:46:32 -0400
Received: from mail-pg1-f194.google.com ([209.85.215.194]:33988 "EHLO
        mail-pg1-f194.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S2387758AbeGWBqc (ORCPT <rfc822;kvm@vger.kernel.org>);
        Sun, 22 Jul 2018 21:46:32 -0400
Received: by mail-pg1-f194.google.com with SMTP id y5-v6so10974646pgv.1;
        Sun, 22 Jul 2018 17:47:56 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=HVOb+vFJlwfEy9YaU1RuD0B68qKfBVjS/jc+VgEZqW0=;
        b=RlgELKSSZvKk8mNvtoULRSysBEm80g7dWK8AaK5v9JZHG0aHusnNw0+ibu2su7ko9N
         dDrS9Sk4bwkRLvw/4iWqnw7ob4ZbitlgfwhQ02Djb6V6SHm5fRU2PHuDjRR4T2+wFg9r
         niIL4EmxVMJAqvFREIOgLecnGDXiweVHc8jZ6Q+8hJ4wyJJijAEDOQYFiDi6OgYh+fvC
         T2Exs5mxIZBoQGhrrjsRmNWt7DxU2YRps5B8w88F7nddg63Uwh+2T8h88LQ1/CDmCub5
         4TBG9VjGPgQy4X/Lyozr7GXRrY5myr1Dq1i0osCz1/GYPCjT6OKdKco9DUjkZZnQI5c1
         pvhQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=HVOb+vFJlwfEy9YaU1RuD0B68qKfBVjS/jc+VgEZqW0=;
        b=jvX7jqqMdAaWxNI0QkuCiKdkKuTLvkqdbGekdURNiiCeORHEXhxRN5HZCG+KGyb7IP
         k32uI9Nodn9eyfhXXtPivTAXzdGiIvkDS4G/PPDyePGDdy+Z+xNYDwBHoton2ZFLVFFM
         TgLkUYoCT7ZTL5i8aberY35F0lig2G7rXUDZYp1JgiBPkz59loXDbdYPXyF4p5Z7Hfth
         nhbKd/cVghJe0DGwUUHhbcbM3q54fqLMeR/WevoHJxLG94R3MPzTksbYCZ57Y581iMbG
         edCEEm+9UOEqLMEIINm8N5HCjDAJHLjeKlgn7motdpB3Ide9MmdIqpviiv8oO6uewdkY
         EPrw==
X-Gm-Message-State: AOUpUlFz5fjjYSDSe8nRzgyh4UibkX6ZmIc7d/69V8/Zn9j1t7HWYl58
        zZgBrCcCDvguj2KTTMSZTPf3EZxd
X-Google-Smtp-Source: 
 AAOMgpd2IISaPVLgsQcw9cQE9Hq7yiV45bGe8ZDguJ1f2hvIYEjnHDx16f8A4RNrGZUyKyFDmN4HHg==
X-Received: by 2002:a63:e914:: with SMTP id
 i20-v6mr10275686pgh.10.1532306876324;
        Sun, 22 Jul 2018 17:47:56 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 e82-v6sm11646069pfk.87.2018.07.22.17.47.54
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Sun, 22 Jul 2018 17:47:55 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Vitaly Kuznetsov <vkuznets@redhat.com>
Subject: [PATCH v4 RESEND 6/6] KVM: X86: Expose PV_SEND_IPI CPUID feature bit
 to guest
Date: Mon, 23 Jul 2018 08:47:39 +0800
Message-Id: <1532306859-31166-7-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1532306859-31166-1-git-send-email-wanpengli@tencent.com>
References: <1532306859-31166-1-git-send-email-wanpengli@tencent.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

Expose PV_SEND_IPI feature bit to guest, the guest can check this feature
bit before using paravirtualized send IPIs.

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Cc: Vitaly Kuznetsov <vkuznets@redhat.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 Documentation/virtual/kvm/cpuid.txt | 4 ++++
 arch/x86/kvm/cpuid.c                | 3 ++-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/Documentation/virtual/kvm/cpuid.txt b/Documentation/virtual/kvm/cpuid.txt
index ab022dc..97ca194 100644
--- a/Documentation/virtual/kvm/cpuid.txt
+++ b/Documentation/virtual/kvm/cpuid.txt
@@ -62,6 +62,10 @@ KVM_FEATURE_ASYNC_PF_VMEXIT        ||    10 || paravirtualized async PF VM exit
                                    ||       || can be enabled by setting bit 2
                                    ||       || when writing to msr 0x4b564d02
 ------------------------------------------------------------------------------
+KVM_FEATURE_PV_SEND_IPI            ||    11 || guest checks this feature bit
+                                   ||       || before using paravirtualized
+                                   ||       || send IPIs.
+------------------------------------------------------------------------------
 KVM_FEATURE_CLOCKSOURCE_STABLE_BIT ||    24 || host will warn if no guest-side
                                    ||       || per-cpu warps are expected in
                                    ||       || kvmclock.
diff --git a/arch/x86/kvm/cpuid.c b/arch/x86/kvm/cpuid.c
index 7e042e3..7bcfa61 100644
--- a/arch/x86/kvm/cpuid.c
+++ b/arch/x86/kvm/cpuid.c
@@ -621,7 +621,8 @@ static inline int __do_cpuid_ent(struct kvm_cpuid_entry2 *entry, u32 function,
 			     (1 << KVM_FEATURE_CLOCKSOURCE_STABLE_BIT) |
 			     (1 << KVM_FEATURE_PV_UNHALT) |
 			     (1 << KVM_FEATURE_PV_TLB_FLUSH) |
-			     (1 << KVM_FEATURE_ASYNC_PF_VMEXIT);
+			     (1 << KVM_FEATURE_ASYNC_PF_VMEXIT) |
+			     (1 << KVM_FEATURE_PV_SEND_IPI);
 
 		if (sched_info_on())
 			entry->eax |= (1 << KVM_FEATURE_STEAL_TIME);
